src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs

```diff
@@ -97,5 +109,103 @@ public static void AddPropertyIfExists(IDictionary<string, string> properties, s
             ProjectStyle.Standalone,
             ProjectStyle.ProjectJson
         };
+
+        public static async Task<bool> RestoreAsync(
+            DependencyGraphSpec dependencyGraphSpec,
+            bool interactive,
+            bool recursive,
+            bool noCache,
+            bool ignoreFailedSources,
+            bool disableParallel,
+            bool force,
+            bool forceEvaluate,
+            bool hideWarningsAndErrors,
+            Common.ILogger log,
+            CancellationToken cancellationToken)
+        {
+            try
+            {
+                DefaultCredentialServiceUtility.SetupDefaultCredentialService(log, !interactive);
+
+                // Set connection limit
+                NetworkProtocolUtility.SetConnectionLimit();
+
+                // Set user agent string used for network calls
+#if IS_CORECLR
+                UserAgent.SetUserAgentString(new UserAgentStringBuilder("NuGet .NET Core MSBuild Task")
+                    .WithOSDescription(RuntimeInformation.OSDescription));
+#else
+                // OS description is set by default on Desktop
+                UserAgent.SetUserAgentString(new UserAgentStringBuilder("NuGet Desktop MSBuild Task"));
+#endif
+
+                // This method has no effect on .NET Core.
+                NetworkProtocolUtility.ConfigureSupportedSslProtocols();
+
+                var providerCache = new RestoreCommandProvidersCache();
+
+                using (var cacheContext = new SourceCacheContext())
+                {
+                    cacheContext.NoCache = noCache;
+                    cacheContext.IgnoreFailedSources = ignoreFailedSources;
+
+                    // Pre-loaded request provider containing the graph file
+                    var providers = new List<IPreLoadedRestoreRequestProvider>();
+
+                    // Add all child projects
+                    if (recursive)
+                    {
+                        AddAllProjectsForRestore(dependencyGraphSpec);
+                    }
+
+                    if (dependencyGraphSpec.Restore.Count < 1)
+                    {
+                        // Restore will fail if given no inputs, but here we should skip it and provide a friendly message.
+                        log.LogMinimal(Strings.NoProjectsToRestore);
+                        return true;
+                    }
```

Comment:
This reverses the order of the [original statements](https://github.com/NuGet/NuGet.Client/pull/3125/files#diff-2e8d36907ec0ce384a96f5fa1b67d718L167-L178).  Does this fix some bug?

Reply:
I think I was thrown off by all of this code:

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs#L59-L114

But I'll fix the order just so I don't cause any issues